onnxruntime/contrib_ops/cpu/transformers/beam_search.cc

```diff
@@ -0,0 +1,650 @@
+  gsl::span<T>& next_token_scores = beam_state.next_token_scores;
+  Status status = SoftmaxCPU<T>(batch_beam_size,  // rows
+                                vocab_size,       // elements per row
+                                input_length > 1 ? next_token_logits.data() : logits_data,
+                                next_token_scores.data(),
+                                true,
+                                thread_pool_);
+  if (!status.IsOK()) {
+    return status;
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpTensor("next_token_scores after softmax", next_token_scores.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  // Apply all score processors that updates scores
+  logits_processors_.Process(&(beam_state.sequences), next_token_scores);
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpTensor("next_token_scores after logits processor", next_token_scores.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  // Add beam score to next token scores. Corresponding python code is like:
+  //    next_token_scores = next_token_scores + beam_scores[:, None].expand_as(next_token_scores)
+  // TODO: use thread pool to parrellel
+  int offset = 0;
+  int batch_beam_index = 0;
+  for (int i = 0; i < parameters_->batch_size; i++) {
+    for (int j = 0; j < parameters_->num_beams; j++, batch_beam_index++) {
+      for (int k = 0; k < parameters_->vocab_size; k++, offset++) {
+        next_token_scores[offset] += beam_state.beam_scores[batch_beam_index];
+      }
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpTensor("next_token_scores after adding beam_scores", next_token_scores.data(), parameters_->batch_size, parameters_->num_beams, vocab_size);
+#endif
+
+  if (parameters_->output_scores) {
+    // Append next token scores to the scores output.
+    gsl::copy(next_token_scores, beam_state.remaining_scores);
+    beam_state.remaining_scores = beam_state.remaining_scores.subspan(next_token_scores.size());
+  }
+
+  // Apply top-k selection like the following:
+  //   next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
+  //   next_token_scores, next_tokens = torch.topk(next_token_scores, 2 * num_beams, dim=1, largest=True, sorted=True)
+  int64_t next_token_scores_dims[] = {parameters_->batch_size, parameters_->num_beams * vocab_size};
+  TensorShape next_token_scores_shape(&next_token_scores_dims[0], 2);
+  auto element_type = DataTypeImpl::GetType<T>();
+  OrtValue next_token_scores_value;
+  Tensor::InitOrtValue(element_type, next_token_scores_shape, next_token_scores.data(), allocator->Info(), next_token_scores_value);
+  const Tensor& input = next_token_scores_value.Get<Tensor>();
+
+  const int axis = 1;
+  const unsigned top_k = static_cast<unsigned>(2 * parameters_->num_beams);
+  const bool largest = true;
+  const bool sorted = true;  // results returned in sorted order.
```

Comment:
Why does it need to be sorted?

Reply:
we need to process best results first (in some case, we can stop early so top_k shall be sorted).

Reply:
I see, why 2*num_beams though?

Reply:
some processor (like vocab mask) will filter some "bad words". That means we shall get top_k > num_beams. 2 * num_beams seems to be a reasonable value chosen in huggingface code that we followed.